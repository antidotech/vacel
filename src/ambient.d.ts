declare type Class<T> = { new (...args: any[]): T }
declare type Emit<T, U> = { [K in Exclude<keyof T, U>]: T[K] }
