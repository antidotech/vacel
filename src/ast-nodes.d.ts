import { Node } from './nodes/node'
import { Token } from './parser/tokenizer'

export type Literal =
  | BooleanLiteral
  | StringLiteral
  | MultilineLiteral
  | DurationLiteral
  | NumericLiteral

export type Expression =
  | Literal
  | Identifier
  | Header
  | Ip
  | Member
  | BooleanExpression
  | UnaryExpression
  | FunCallExpression
  | ConcatExpression
  | BinaryExpression
  | LogicalExpression

export type Statement =
  | ExpressionStatement
  | IncludeStatement
  | ImportStatement
  | CallStatement
  | DeclareStatement
  | AddStatement
  | SetStatement
  | UnsetStatement
  | ReturnStatement
  | ErrorStatement
  | RestartStatement
  | SyntheticStatement
  | LogStatement
  | IfStatement
  | SubroutineStatement
  | AclStatement
  | CustomStatement
  | BackendStatement
  | TableStatement

export interface File extends Node {
  type: 'File'
  filePath: string
  program: Program
  comments: Array<Token>
}

export interface Program extends Node {
  type: 'Program'
  body: Array<Statement | Comment>
}

export interface Comment extends Node {
  type: 'Comment'
  // TODO: add `type` prop
  body: string
}

// TODO: remove this node
export interface Block extends Node {
  type: 'Block'
  body: Array<Statement | Comment>
}

export interface BooleanLiteral extends Node {
  type: 'BooleanLiteral'
  value: string
}

export interface StringLiteral extends Node {
  type: 'StringLiteral'
  value: string
}

// TODO: merge with StringLiteral
export interface MultilineLiteral extends Node {
  type: 'MultilineLiteral'
  value: string
}

export interface DurationLiteral extends Node {
  type: 'DurationLiteral'
  value: string
}

export interface NumericLiteral extends Node {
  type: 'NumericLiteral'
  value: string
}

export interface Identifier extends Node {
  type: 'Identifier'
  name: string
}

export interface Header extends Node {
  type: 'Header'
  name: string
}

export interface Ip extends Node {
  type: 'Ip'
  value: string
  cidr?: number
}

export interface Member extends Node {
  type: 'Member'
  base: Identifier | Member
  member: Identifier | Header
}

export interface BooleanExpression extends Node {
  type: 'BooleanExpression'
  body: Expression
}

export interface UnaryExpression extends Node {
  type: 'UnaryExpression'
  operator: string
  argument: Expression
}

export interface FunCallExpression extends Node {
  type: 'FunCallExpression'
  callee: Member | Identifier
  arguments: Array<Expression>
}

export interface ConcatExpression extends Node {
  type: 'ConcatExpression'
  body: Array<Expression>
}

export interface BinaryExpression extends Node {
  type: 'BinaryExpression'
  left: Expression
  right: Expression
  operator: string
}

export interface LogicalExpression extends Node {
  type: 'LogicalExpression'
  left: Expression
  right: Expression
  operator: string
}

export interface ExpressionStatement extends Node {
  type: 'ExpressionStatement'
  body: Expression
}

export interface IncludeStatement extends Node {
  type: 'IncludeStatement'
  module: StringLiteral
}
export interface ImportStatement extends Node {
  type: 'ImportStatement'
  module: Identifier
}

export interface CallStatement extends Node {
  type: 'CallStatement'
  subroutine: Identifier
}

export interface DeclareStatement extends Node {
  type: 'DeclareStatement'
  id: Identifier | Member
  valueType: 'STRING' | 'BOOL' | 'BOOLEAN' | 'INTEGER' | 'FLOAT'
}

export interface AddStatement extends Node {
  type: 'AddStatement'
  left: Identifier | Member
  right: Expression
  operator: string
}

export interface SetStatement extends Node {
  type: 'SetStatement'
  left: Identifier | Member
  right: Expression
  operator: string
}

export interface UnsetStatement extends Node {
  type: 'UnsetStatement'
  id: Identifier | Member
}

export interface ReturnStatement extends Node {
  type: 'ReturnStatement'
  action: 'pass' | 'hit_for_pass' | 'lookup' | 'pipe' | 'deliver'
}

export interface ErrorStatement extends Node {
  type: 'ErrorStatement'
  status: Literal
  message?: Expression
}

export interface RestartStatement extends Node {
  type: 'RestartStatement'
}

export interface SyntheticStatement extends Node {
  type: 'SyntheticStatement'
  response: Expression
}

export interface LogStatement extends Node {
  type: 'LogStatement'
  content: Expression
}

export interface IfStatement extends Node {
  type: 'IfStatement'
  test: Expression
  consequent: Array<Statement>
  alternative?: IfStatement | Array<Statement>
}

export interface SubroutineStatement extends Node {
  type: 'SubroutineStatement'
  id: Identifier
  body: Array<Statement>
}

export interface AclStatement extends Node {
  type: 'AclStatement'
  id: Identifier
  body: Array<Ip>
}

export interface CustomStatement extends Node {
  type: 'CustomStatement'
  directive: string
  body: Array<Node>
}

export type BackendDef = { key: string; value: Expression | Array<BackendDef> }

export interface BackendStatement extends Node {
  type: 'BackendStatement'
  id: Identifier
  body: Array<BackendDef>
}

export type TableDef = { key: string; value: string }

export interface TableStatement extends Node {
  type: 'TableStatement'
  id: Identifier
  body: Array<TableDef>
}
