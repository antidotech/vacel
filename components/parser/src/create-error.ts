import { Position } from '@vaceline/types'
import * as chalk from '@vaceline/utils'

const MARGIN = 2
const HORIZONTAL_MARK = '> '
const VERTICAL_MARK = '^'

export function createError(
  source: string,
  message: string,
  start: Position,
  end: Position
): SyntaxError {
  const topLineNumber = start.line >= MARGIN ? start.line - MARGIN : start.line
  const topLineOffset = topLineNumber - 1 // zero-offset

  const indicator = source
    .split('\n')
    .slice(topLineOffset, topLineOffset + MARGIN * 2 + 1)
    .map((line, i) => {
      const lineNumber = topLineNumber + i
      const lineOffset = topLineOffset + i

      const bar = chalk.gray(String(lineNumber) + ' | ')

      if (lineNumber === start.line) {
        return [
          chalk.red(HORIZONTAL_MARK) + bar + line,
          ' '.repeat(5 + String(lineOffset).length + start.column - 1) +
            chalk.red(VERTICAL_MARK.repeat(end.offset + 1 - start.offset)),
        ]
      }

      return ' '.repeat(HORIZONTAL_MARK.length) + bar + line
    })
    .flat()
    .join('\n')

  return new SyntaxError(message + '\n\n' + indicator + '\n')
}
