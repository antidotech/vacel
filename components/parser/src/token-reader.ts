import { Comment } from '@vaceline/types'
import { Token } from './tokenizer'

export class TokenReader {
  private tokens: Array<Token>
  private cur: number

  comments: Array<Comment>

  constructor(tokens: Array<Token>) {
    this.tokens = tokens
    this.comments = []

    this.cur = 0
  }

  getCurrentToken(): Token {
    return this.tokens[this.cur - 1]
  }

  jumpTo(cur: number) {
    this.cur = cur
  }

  getCursor(): number {
    return this.cur
  }

  getToken(cur: number): Token | null {
    return this.tokens[cur]
  }

  read(): Token {
    const token = this.tokens[this.cur++]

    if (!token) {
      throw new SyntaxError('Unexpected EOF')
    }

    if (token.type === 'comment') {
      this.comments.push({
        type: 'CommentLine',
        value: token.value,
        loc: token.loc,
      })
      return this.read()
    }

    return token
  }

  peek(skip = 0): Token | undefined {
    const token = this.tokens[this.cur + skip]

    if (!token) return undefined

    if (token.type === 'comment') {
      return this.peek(skip + 1)
    }

    return token
  }

  // TODO: This can be replaced with read()
  take(): void {
    this.read()
  }
}
