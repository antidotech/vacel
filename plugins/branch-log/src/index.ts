import { Node } from '@vaceline/types'
import { traverse } from '@vaceline/traverser'

// req.http.Vaceline-Branch-Log
const varBranchLog: Node = {
  type: 'Member',
  base: {
    type: 'Member',
    base: { type: 'Identifier', name: 'req' },
    member: { type: 'Identifier', name: 'http' },
  },
  member: { type: 'Identifier', name: 'Vaceline-Branch-Log' },
}

const getLoc = (node: Node) =>
  node.loc ? `${node.loc.start.line}:${node.loc.start.column}` : 'synthethic'

export default (ast: Node) => {
  traverse(ast, {
    entry({ node }) {
      if (node.type === 'SubroutineStatement') {
        if (node.id.name === 'vcl_recv') {
          node.body.unshift(
            /**
             * set req.http.Vaceline-Branch-Log = "(vcl_recv)${line}:${col}";
             */
            {
              type: 'SetStatement',
              left: varBranchLog,
              right: {
                type: 'StringLiteral',
                value: `"(${node.id.name})${getLoc(node)}"`,
              },
              operator: '=',
            }
          )
        } else {
          node.body.unshift(
            /**
             * add req.http.Vaceline-Branch-Log = "(${nodeName})${line}:${col}";
             */
            {
              type: 'AddStatement',
              left: varBranchLog,
              right: {
                type: 'StringLiteral',
                value: `"(${node.id.name})${getLoc(node)}"`,
              },
              operator: '=',
            }
          )
        }

        if (node.id.name === 'vcl_deliver') {
          node.body.push(
            /**
             * set resp.http.Vaceline-Branch-Log = std.collect(
             *   req.http.Vaceline-Branch-Log
             * );
             */
            {
              type: 'ExpressionStatement',
              body: {
                type: 'FunCallExpression',
                callee: {
                  type: 'Member',
                  base: {
                    type: 'Identifier',
                    name: 'std',
                  },
                  member: { type: 'Identifier', name: 'collect' },
                },
                args: [varBranchLog],
              },
            },
            {
              type: 'SetStatement',
              left: {
                type: 'Member',
                base: {
                  type: 'Member',
                  base: { type: 'Identifier', name: 'resp' },
                  member: { type: 'Identifier', name: 'http' },
                },
                member: { type: 'Identifier', name: 'Vaceline-Branch-Log' },
              },
              right: {
                type: 'Member',
                base: {
                  type: 'Member',
                  base: { type: 'Identifier', name: 'req' },
                  member: { type: 'Identifier', name: 'http' },
                },
                member: { type: 'Identifier', name: 'Vaceline-Branch-Log' },
              },
              operator: '=',
            }
          )
        }
      } else if (node.type === 'IfStatement') {
        node.consequent.unshift(
          /**
           * add req.http.Vaceline-Branch-Log = "${line}:${col}";
           */
          {
            type: 'AddStatement',
            left: varBranchLog,
            right: {
              type: 'StringLiteral',
              value: node.loc
                ? `"(anonymous)${node.loc.start.line}:${node.loc.start.column}"`
                : '"synthethic"',
            },
            operator: '=',
          }
        )
      }
    },
  })
}
