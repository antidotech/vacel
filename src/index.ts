import { resolve as resolvePath } from 'path'
import { existsSync, readFileSync } from 'fs'
import assert from 'assert'

import {
  parse,
  generate,
  // traverse,
} from './lib'
import { d } from './nodes'
import { GenerateOptions } from './generator'

interface TransformResult {
  code: string
  map: string
  ast: d.Program
}

export { parse, generate as transformFromAst }

export type Options = {} & GenerateOptions

export const transform = (
  code: string,
  options: Partial<Options> = {}
): TransformResult => {
  const ast = parse(code)

  const result: Partial<TransformResult> = generate(ast, options)

  result.ast = ast

  return result as TransformResult
}

export const transformFile = (
  filePath: string,
  options: Partial<Options> = {}
): TransformResult => {
  const inputPath = resolvePath(filePath)
  assert(existsSync(inputPath), 'File not found: ' + inputPath)

  const ast = parse(readFileSync(filePath, 'utf8'))
  const result: Partial<TransformResult> = generate(ast, options)

  result.ast = ast

  return result as TransformResult
}
