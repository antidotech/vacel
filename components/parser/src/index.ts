import { Tokenizer, Token, TokenType, isToken } from './tokenizer'
import {
  Node,
  Located,
  NodeType,
  Statement,
  Program,
  Position,
  Comment,
} from '@vaceline/types'
import { createError } from './create-error'
import { TokenReader } from './token-reader'

import { buildDebug } from '@vaceline/utils'
import { parseStmt } from './statement/index'
import { parseCompound } from './compound'

const debug = {
  start: buildDebug('parser:start'),
  finish: buildDebug('parser:finish'),
  trace: buildDebug('parser:trace'),
}

export const parse = (source: string): Program => new Parser(source).parse()

interface ParsingState {
  token: Token
  pos: Position
  comments?: Array<Comment>
}

export class Parser {
  source: string
  private reader: TokenReader

  constructor(source: string /* opts: { keywords?: Array<string> } = {} */) {
    const tokens = new Tokenizer(source).tokenize()

    this.source = source
    this.reader = new TokenReader(tokens)
  }

  parse(): Located<Program> {
    const body = parseCompound<Statement>(this, parseStmt)

    const pos = { offset: 0, line: 1, column: 1 }

    const node: Located<Program> = {
      type: 'Program',
      body,
      loc: {
        start: pos,
        end: this.reader.getCurrentToken().loc.end,
      },
    }

    node.leadingComments = []
    node.innerComments = this.parseInnerComments()
    node.trailingComments = this.parseTrailingComments()

    return node
  }

  parseNode<T extends NodeType>(
    token: Token,
    parse: (state: ParsingState) => Node & { type: T }
  ): Located<Node & { type: T }> {
    debug.start(token)

    const state: ParsingState = {
      token,
      pos: token.loc.start,
    }

    const leadingComments = this.parseLeadingComments(state.pos)

    const node = parse(state)

    if (!node.loc) {
      node.loc = {
        start: state.pos,
        end: this.reader.getCurrentToken().loc.end,
      }
    }

    node.leadingComments = leadingComments
    node.innerComments = this.parseInnerComments()
    node.trailingComments = this.parseTrailingComments()

    debug.finish(node)

    // @ts-expect-error FIXME:
    return node
  }

  parseLeadingComments(pos: Position): Array<Comment> {
    const leadingComments = []

    let i = this.reader.comments.length - 1

    while (this.reader.comments[i]?.loc?.start.line ?? Infinity < pos.line) {
      leadingComments?.push(this.reader.comments[i])
      i--
    }

    this.reader.comments = this.reader.comments.slice(0, i + 1)

    return leadingComments.reverse()
  }

  parseInnerComments(): Array<Comment> {
    const innerComments = this.reader.comments.slice()

    this.reader.comments = []

    return innerComments
  }

  parseTrailingComments(): Array<Comment> {
    const trailingComments: Array<Comment> = []

    let cur = this.reader.getCursor()
    let token = this.reader.getToken(cur)

    while (
      token?.type === 'comment' &&
      token.loc.start.line === this.reader.getCurrentToken().loc.end.line
    ) {
      this.reader.jumpTo(cur + 1)

      trailingComments.push({
        type: 'CommentLine',
        value: token.value,
        loc: token.loc,
      })

      token = this.reader.getToken(++cur)
    }

    return trailingComments
  }

  read() {
    return this.reader.read()
  }
  peek() {
    return this.reader.peek()
  }
  take() {
    return this.reader.take()
  }
  getCursor() {
    return this.reader.getCursor()
  }
  jumpTo(cur: number) {
    return this.reader.jumpTo(cur)
  }
  getCurrentToken(): Token {
    return this.reader.getCurrentToken()
  }

  validateNode<T extends Array<NodeType>>(
    node: Located<Node>,
    ...types: T
  ): Located<Node & { type: T[number] }> {
    if (!types.includes(node.type)) {
      throw createError(
        this.source,
        'Expected one of [' + types.join(', ') + ']',
        node.loc.start,
        node.loc.end
      )
    }

    return node as Located<Node & { type: T }>
  }

  validateToken<T extends TokenType, U extends string>(
    token: Token,
    type: T,
    value?: U
  ): Token & { type: T; value: U } {
    if (!isToken(token, type, value)) {
      throw createError(
        this.source,
        `Expected '${value}' ${type} token`,
        token.loc.start,
        token.loc.end
      )
    }

    return token as Token & { type: T; value: U }
  }
}
