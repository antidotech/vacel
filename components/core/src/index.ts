import { resolve as resolvePath } from 'path'
import { existsSync, readFileSync } from 'fs'
import assert from 'assert'

import { parse } from '@vaceline/parser'
import { generate, GenerateOptions } from '@vaceline/generator'
import { Program } from '@vaceline/types'

interface TransformResult {
  code: string
  map: string
  ast: Program
}

export { parse, generate as transformFromAst }

export type Options = {} & GenerateOptions

export function transform(
  code: string,
  options: Partial<Options> = {}
): TransformResult {
  const ast = parse(code)

  const result: Partial<TransformResult> = generate(ast, options)

  result.ast = ast

  return result as TransformResult
}

export function transformFile(
  filePath: string,
  options: Partial<Options> = {}
): TransformResult {
  const inputPath = resolvePath(filePath)
  assert(existsSync(inputPath), 'File not found: ' + inputPath)

  const ast = parse(readFileSync(filePath, 'utf8'))
  const result: Partial<TransformResult> = generate(ast, options)

  result.ast = ast

  return result as TransformResult
}
