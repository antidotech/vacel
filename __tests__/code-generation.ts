import { wrap } from 'jest-snapshot-serializer-raw'
import { testdocs } from './helpers/snapshot'
import { generate } from '../src'

for (const testdoc of testdocs) {
  for (const testcase of testdoc.basic) {
    describe(`(${testdoc.name})`, () => {
      it(testcase.name, () => {
        const ast = testdoc.parse.call(testdoc.parse, testcase.code)
        const { code } = generate(ast)

        expect(wrap(code)).toMatchSnapshot()
      })
    })
  }
}
