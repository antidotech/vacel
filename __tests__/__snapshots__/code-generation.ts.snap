// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(AclStatement) basic 1`] = `
aclaaa127.0.0.1
localhost;
`;

exports[`(AddStatement) TODO: cover operand types
addition 1`] = `add var.vacel += "vacel";`;

exports[`(AddStatement) assignment 1`] = `add var.vacel = "vacel";`;

exports[`(AddStatement) division 1`] = `add var.vacel *= "vacel";`;

exports[`(AddStatement) multiplication 1`] = `add var.vacel *= "vacel";`;

exports[`(AddStatement) substraction 1`] = `add var.vacel -= "vacel";`;

exports[`(BackendStatement) basic 1`] = `
backend vaceline_backend {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "vaceline.io";
  .first_byte_timeout = 5s;
  .max_connections = 100;
  .between_bytes_timeout = 10s;
  .share_key = "f8585BOxnGQDMbnkJoM1e";
  .ssl = true;
  .probe = {
    .request = "GET /healthcheck HTTP/1.1"
    "Host: vaceline.io"
    "Connection: close"
    "User-Agent: Vaceline healthcheck";
    .threshold = 1;
    .window = 2;
    .timeout = 10s;
    .initial = 1;
    .expected_response = 200;
    .interval = 60s;
  }
}
`;

exports[`(BinaryExpression) combinations 1`] = `req.http.Debug == "1" == "2" == "3"`;

exports[`(BinaryExpression) equal 1`] = `req.http.Debug == "1"`;

exports[`(BinaryExpression) match 1`] = `req.http.Debug ~ "1"`;

exports[`(BinaryExpression) not equal 1`] = `req.http.Debug != "1"`;

exports[`(BooleanExpression) assign 1`] = `
set
var.Debug
`;

exports[`(BooleanExpression) with binary 1`] = `
(
  req.http.Debug == "1"
)
`;

exports[`(BooleanExpression) with logical 1`] = `
(
  var.Debug &&
  "1"
)
`;

exports[`(BooleanLiteral) false 1`] = `false`;

exports[`(BooleanLiteral) true 1`] = `true`;

exports[`(CallStatement) basic 1`] = `call vaceline_func;`;

exports[`(ConcatExpression) expressions 1`] = `
"a"
1
req.http.A
`;

exports[`(ConcatExpression) strings 1`] = `
"a"
"b"
`;

exports[`(DeclareStatement) bool 1`] = `declare local var.Vaceline BOOL;`;

exports[`(DeclareStatement) float 1`] = `declare local var.Vaceline FLOAT;`;

exports[`(DeclareStatement) integer 1`] = `declare local var.Vaceline INTEGER;`;

exports[`(DeclareStatement) string 1`] = `declare local var.Vaceline STRING;`;

exports[`(DurationLiteral) day 1`] = `1d`;

exports[`(DurationLiteral) min 1`] = `1m`;

exports[`(DurationLiteral) month 1`] = `1`;

exports[`(DurationLiteral) ms 1`] = `1ms`;

exports[`(DurationLiteral) s 1`] = `1s`;

exports[`(DurationLiteral) year 1`] = `1`;

exports[`(ErrorStatement) basic 1`] = `error404;`;

exports[`(ErrorStatement) with response 1`] = `error404"Not Found";`;

exports[`(ErrorStatement) with variable response 1`] = `error800req.http.Debug;`;

exports[`(ExpressionStatement) with FunCallExpression 1`] = `
log(
  a
);
`;

exports[`(ExpressionStatement) with Identifer 1`] = `vaceline;`;

exports[`(ExpressionStatement) with Literal 1`] = `"string";`;

exports[`(ExpressionStatement) with Member & Identifier 1`] = `req.http.Vaceline;`;

exports[`(File) file 1`] = `"file";`;

exports[`(FunCallExpression) \`if\` func 1`] = `
if(
"a", "b", "c")
`;

exports[`(FunCallExpression) normal 1`] = `
func(
"a")
`;

exports[`(FunCallExpression) with multiple args 1`] = `
func(
"a", "b")
`;

exports[`(FunCallExpression) with whitespace 1`] = `
func(
"a")
`;

exports[`(Identifier) normal 1`] = `vaceline`;

exports[`(IfStatement) basic 1`] = `
if (var.condition) {
  set var.value = 1;
}
`;

exports[`(IfStatement) with else 1`] = `
if (var.condition) {
  set var.value = 1;
}elseset var.value = 2;
`;

exports[`(IfStatement) with else if 1`] = `
if (var.condition) {
  set var.value = 1;
}else{if (!var.condition) {
  set var.value = 2;
}elseset var.value = 0;}
`;

exports[`(ImportStatement) basic 1`] = `module;`;

exports[`(IncludeStatement) basic 1`] = `include "external_library";`;

exports[`(Ip) 6to4 mapping 1`] = `"2002:c000:0204::"`;

exports[`(Ip) 6to4 mapping 2`] = `"::FFFF:192.0.2.4"`;

exports[`(Ip) 6to4 mapping 3`] = `"::1"`;

exports[`(Ip) IPv4 1`] = `"192.0.2.0"`;

exports[`(Ip) IPv4 with cidr 1`] = `"192.0.2.0"`;

exports[`(Ip) IPv6 1`] = `"2001:db8::1"`;

exports[`(Ip) IPv6 with cidr 1`] = `"2001:db8::1"`;

exports[`(Ip) localhost 1`] = `"localhost"`;

exports[`(Ip) unspecified address 1`] = `"::"`;

exports[`(LogStatement) basic 1`] = `
log"log"
req.service_id
"vaceline-log :: "
{"	timestamp_us:"}
time.start.usec
{"	host:"}
regsuball(
req.http.X-Forwarded-Host, {"	"}, "");
`;

exports[`(LogicalExpression) and 1`] = `
req.http.Debug &&
"1"
`;

exports[`(LogicalExpression) combination 1`] = `
req.http.Debug ||
"1" &&
"2"
`;

exports[`(LogicalExpression) or 1`] = `
req.http.Debug ||
"1"
`;

exports[`(Member) basic 1`] = `req.http.Vaceline`;

exports[`(MultilineLiteral) basic 1`] = `
{"
  vaceline
"}
`;

exports[`(NumericLiteral) integer 1`] = `10`;

exports[`(NumericLiteral) zero 1`] = `0`;

exports[`(Program) single handling 1`] = `
sub vcl_recv {
  declare local var.vacelinDebug STRING;
  if (req.http.Vacelin:Some == "Some") {
    add var.vacelinDebug = "Vacelin Debug Log";
  }
}
`;

exports[`(RestartStatement) basic 1`] = `restart;`;

exports[`(ReturnStatement) with parens 1`] = `returnpass;`;

exports[`(ReturnStatement) without parens 1`] = `returnpass;`;

exports[`(SetStatement) addition 1`] = `set var.vacel += "vacel";`;

exports[`(SetStatement) assignment 1`] = `set var.vacel = "vacel";`;

exports[`(SetStatement) division 1`] = `set var.vacel *= "vacel";`;

exports[`(SetStatement) multiplication 1`] = `set var.vacel *= "vacel";`;

exports[`(SetStatement) substraction 1`] = `set var.vacel -= "vacel";`;

exports[`(StringLiteral) basic 1`] = `"string"`;

exports[`(StringLiteral) not a comment 1`] = `"# not a comment"`;

exports[`(StringLiteral) not an inline comment 1`] = `"/* not an inline comment */"`;

exports[`(SubroutineStatement) basic 1`] = `
sub vcl_recv {
  set vaceline = "aaa";
}
`;

exports[`(SyntheticStatement) basic 1`] = `synthetic"";`;

exports[`(TableStatement) single key-value 1`] = `tablevaceline_table"key":"value"`;

exports[`(TableStatement) single key-values 1`] = `
tablevaceline_table"key":"value"
"key2":"value2"
`;

exports[`(TableStatement) with trailing comma 1`] = `
tablevaceline_table"key":"value"
"key2":"value2"
`;

exports[`(UnaryExpression) boolean literal 1`] = `!true`;

exports[`(UnaryExpression) variable 1`] = `!condition`;

exports[`(UnsetStatement) member 1`] = `unset var.value;`;

exports[`(UnsetStatement) variable 1`] = `unset value;`;

exports[`(ValuePair) basic 1`] = `Cookie:Vacelin`;
